### ISSUE TITLE
typeof operator used on an inferred record type produces a cryptic error

### ISSUE THREAD

**Issue by fingerartur** [google/closure-compiler#4024]
As described [here](https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System), the `typeof` operator is probably not intended to be used on inferred types. However, in Typescript it is possible to do it, and it is in fact a very common use case.

Here is an example of using the typeof operator on an inferred record type.
```js
const x = {
  name: "Bat",
  surname: "Man"
}

/**
 * @type {typeof x} // This raises cryptic error
 */
const y = {
  name: "Cat",
  surname: "Woman"
}
```

And it produces this error:
```
./src/js/index.js:10:12: WARNING - [JSC_TYPE_MISMATCH] initializing variable
found   : {name: string, surname: string}
required: {name: string, surname: string}
  10|   const y = {
                  ^
  11|     name: "Cat",
      ^^^^^^^^^^^^^^^^
  12|     surname: "Woman"
      ^^^^^^^^^^^^^^^^^^^^
  13|   }
      ^^^
```
This error message does not make a lot of sense. Two identical types do not match? A better readable error message would be along the lines of "It is forbidden to use typeof operator on inferred types / inferred record types".

Compiler Version: v20221102

Build command:
```
java -jar ./scripts/closureCompiler.jar \
  --entry_point=./src/js/index.js \
  --js=./src/**.js \
  --dependency_mode=PRUNE \
  --warning_level=VERBOSE \
  --js_output_file=./dist/bundle.js \
  --module_resolution=WEBPACK \
  --compilation_level=ADVANCED \
  --jscomp_error=checkDebuggerStatement \
  --jscomp_error=unusedLocalVariables \
  --jscomp_error=reportUnknownTypes;
```

**Comment by brad4d** [comment]
closure-compiler does not do structural matching of types to determine their equality unless they are `@record` types.

```js
// closure-compiler will construct an ad-hoc type that represents the type of this value.
// It won't get a name, and it will only appear directly in the AST attached to the node(s) referring to this value.
const x = {
  name: "Bat",
  surname: "Man"
}

// closure-compiler will generate a distinct ad-hoc type that represents the value of `y`.
// It will then check to see if that type can be considered an instance of the ad-hoc type created for `x`.
// The answer to that question is "no", because the compiler does not do structural matching of types unless
// one or both of them are declared with `@record`.
/**
 * @type {typeof x}
 */
const y = {
  name: "Cat",
  surname: "Woman"
}
```

If you want structural matching, you have to explicitly ask the compiler to do it by declaring an `@record` type.

```js
/** @record */
const Xtype = class {
  constructor() {
    // Declare the fields in the constructor like this.
    /** @type {string} */
    this.name;
    /** @type {string} */
    this.surname;
  }
}

// Structural type matching will recognize that the value of `y` can be used where an `Xtype` is needed.
/**
 * @type {!Xtype}
 */
const y = {
  name: "Cat",
  surname: "Woman"
}
```

It's unfortunate that the error message is confusing, of course,
but it's also not something that comes up much in normal usage.

We might be willing to accept a PR that clarifies the error message,
but we're not likely to change the basic behavior.
One possibility would be to print something like "Anonymous type (<hashcode>) <type description>" in the error message so the user can see that although the contents of the types match, they aren't actually the same.
